### PR TITLE
rpcserver: Refactor and updated documents

### PR DESCRIPTION
**Pull request by JoeGruffins**
For createrawsstx explicitly state that all values are in atoms in help
messages and the documentation.

For createrawssrtx explicitly state that all values are in coins in help
messages and the documentation. Remove range loop for only one input.

For createrawtransaction explicitly state that all values are in coins in
help messages and the documentation. Move conversion to atoms above the
check for max amount and check after converting from coins.

Remove bad white spaces from JSON api documentation.

### CODE REVIEW

**Comment by JoeGruffins** [comment]
Will put that commit message in present tense...

**Comment by JoeGruffins** [comment]
I changed a failure case for createrawtransaction so I thought I should update the test for it.  There isn't one right? Added a new test file for the test as rpcserver looks reserved for live tests. If this is ok, I would love to add more tests to the file.

It seems that only createrawsstx takes everything in atoms, the other two in coins.

I also think it's odd that createrawssrtx takes a flat fee as an argument and not a rate. To find the size of the revocation, although they are pretty much the same, you first have to do half the work of making the raw transaction anyway.

The other two also look like you need to do that precalculation. I can see why noone uses these.

File: rpcserverhandlers_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+)
+
+// testCfg holds the cfg for the test server.
+var testCfg = rpcserverConfig{ChainParams: chaincfg.MainNetParams()}
+
+// testServer is a rpcServer stub used for testing handlers.
+var testServer = &rpcServer{cfg: testCfg}
+
+func TestHandleCreateRawTransaction(t *testing.T) {
+	type test struct {
```
Comment (davecgh) [inline rpcserverhandlers_test.go]: No need for a separate type and I'd prefer it to also be more consistent with the rest of tests in terms of minimizing indententation like so:

```Go
tests := []struct{
	cmd          *types.CreateRawTransactionCmd
	name, result string
	wantErr      bool
	errCode      dcrjson.RPCErrorCode
}{{
	name: "foo",
	// other fields...
}, {
	name: "foo2",
	// other fields...
}}
```

File: rpcserverhandlers_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		cmd          *types.CreateRawTransactionCmd
+		name, result string
+		wantErr      bool
+		errCode      dcrjson.RPCErrorCode
+	}
+	intPtr := func(i int64) *int64 {
```
Comment (davecgh) [inline rpcserverhandlers_test.go]: Not needed.  There is `dcrjson.Int` for this exact purpose.

File: docs/json_rpc_api.mediawiki [hunk 3]
```diff
@@ -450,14 +450,14 @@ the method name for further details such as parameter and return information.
 # <code>inputs</code>: <code>(json array, required)</code> The inputs to the transaction of type.
 #: <code>txid</code>: <code>(string)</code> Unspent tx output hash.
 #: <code>vout</code>: <code><numeric></code> Index of the output being redeemed.
-#: <code>amt</code>: <code><numeric></code> Amount of utxo.
+#: <code>amt</code>: <code><numeric></code> Amount of utxo in atoms.
 #: <code>tree</code>: <code><numeric></code> Which tree utxo is located.
-# <code>amount</code>: <code>(json object, required)</code> json object with the destination addresses as keys and amounts as values.
-#: <code>{"desaddr":"amount", ...}</code>
+# <code>amount</code>: <code>(json object, required)</code> json object with the destination address as key and amount as value in atoms.
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply docs/json_rpc_api.mediawiki]: It looks like having more than one address will cause an error:
https://github.com/decred/dcrd/blob/7feb3715011217a78caffb5988ab9cd430abd129/rpcserver.go#L718

[the remark this answers is not in the document]
Comment (davecgh) [reply docs/json_rpc_api.mediawiki]: You're right.  I misread this as being in the entry for `createrawtransaction`.

File: docs/json_rpc_api.mediawiki [hunk 4]
```diff
@@ -450,14 +450,14 @@ the method name for further details such as parameter and return information.
...
+#: <code>{"desaddr":"amount"}</code>
```
Comment (davecgh) [inline docs/json_rpc_api.mediawiki]: I realize it already said `desaddr`, but what do you think about making it match the CLI help output of `address` or `addr`?

```bash
$ dcrctl help createrawsstx
createrawsstx [{"txid":"value","vout":n,"tree":n,"amt":n},...] amount [{"addr":"value","commitamt":n,"changeaddr":"value","changeamt":n},...]

...
Arguments:
...
2. amount (object, required) JSON object with the destination addresses as keys and amounts as values
{
 "address": n.nnn, (object) The destination address as the key and the amount in DCR as the value
 ...
}
...

Result:
"value" (string) Hex-encoded bytes of the serialized transaction
```

**Comment by davecgh** [comment]
I tested all of the RPC server help updates and can confirm they are displaying properly.

File: docs/json_rpc_api.mediawiki (hunk 4 above)
Comment (JoeGruffins) [reply docs/json_rpc_api.mediawiki]: Also added the map breakdown to the first line after calling `help createrawsstx`

**Comment by JoeGruffins** [comment]
createrawssrtx also only accepts one input, so I made it singular in two of the three place it shows up. However, the middle of the help message still shows it as accepting more than one input. Changing that would take an exception in jsonrpc.reflectTypeToJSONExample, which is crazy since the real answer is to not have it as an array, which would be a breaking change so just leaving it.

**Comment by davecgh** [comment]
Looks good with the latest updates.  Please squash the latest commit and I'll get it merged.

**Comment by davecgh** [comment]
> There isn't one right? Added a new test file for the test as rpcserver_test.go looks reserved for live tests. If this is ok, I would love to add more tests to the file.

There are tests to ensure the proper encoding and decoding of the messages that live alongside the types themselves (e.g. [chainsvrcmds_test.go](https://github.com/decred/dcrd/blob/7feb3715011217a78caffb5988ab9cd430abd129/rpc/jsonrpc/types/chainsvrcmds_test.go)).

> It seems that only createrawsstx takes everything in atoms, the other two in coins.

These inconsistencies are indeed something I am not a fan of, but I worry that changing it, especially in the direction of atoms to coins, can lead to a very bad result for anything using it because it could silently start creating transactions that are for huge amount of coins when it was just mere atoms before.

To be perfectly honest, I wouldn't be horribly against creating an issue for discussing the possibility of deprecating them for eventual removal and perhaps replacing them with differently named versions that use more familiar terminology with the transaction's purpose like `createrawtickettx`, `createrawrevocationtx`.  That would allow the new versions to start with a clean slate without risking the aforementioned issues with potentially silently-induced issues related to amounts.

> I also think it's odd that createrawssrtx takes a flat fee as an argument and not a rate. To find the size of the revocation, although they are pretty much the same, you first have to do half the work of making the raw transaction anyway.

Fair point.  I suspect a better way in general is to use something like `fundtransaction` for this purpose.

**Pull request merged**